File: hackerearth/connected_components.cpp
#include<bits/stdc++.h>
using namespace std;

void dfs(int node, unordered_map<int,vector<int>> &graph,vector<int> &visited){
    if(visited[node]==1) return;
    visited[node] = 1;
    for(int i: graph[node]){
        dfs(i,graph,visited);
    }
}

int main(){
    int n,e;
    cin >> n >> e;
    vector<int> visited(n+1,0);
    unordered_map<int,vector<int>> graph;
    for(int i = 1; i <= n; i++){
        graph[i] = {};
    }
    for(int i = 0; i < e; i++){
        int u,v;
        cin >> u >> v;
        graph[u].emplace_back(v);
        graph[v].emplace_back(u);
    }
    int components = 0;
    for(int i = 1; i <=n; i++){
        if(visited[i]==0){
            components++;
            dfs(i,graph,visited);
        }
    }
    cout << components << endl;
}

File: hackerearth/connected_components.py
def dfs(i):
    if visited[i]:
        return 
    visited[i] = 1 
    for nbr in graph[i]:
        dfs(nbr)

n,e = map(int,input().split())
graph = dict()

visited = [0 for i in range(n+1)]
for i in range(1,n+1):
    graph[i] = []

for i in range(e):
    u,v = map(int,input().split())
    graph[u].append(v)
    graph[v].append(u)

components = 1 
for i in range(1,n+1):
    if not visited[i]:
        components += 1 
        dfs(i)

print(components)
